### ISSUE TITLE
haskell-ci's use of `cabal new-sdist all` can trigger a Cabal bug

### ISSUE THREAD

**Issue by RyanGlScott** [haskell-CI/haskell-ci#186]
I noticed this recently in a [Travis run](https://travis-ci.org/ku-fpg/blank-canvas/jobs/440722473#L1546) for `blank-canvas`:

```
$ cabal new-sdist all
HEAD is now at 3761a8a... Travis: Fully enable GHC 8.6.1
Wrote tarball sdist to
/home/travis/build/ku-fpg/blank-canvas/dist-newstyle/sdist/blank-canvas-0.7.tar.gz
Wrote tarball sdist to
/home/travis/build/ku-fpg/blank-canvas/dist-newstyle/sdist/blank-canvas-examples-0.7.tar.gz
cabal: The impossible happened: a local package isn't local
```

Why did this happen? `blank-canvas`'s `cabal.project` file contains a `source-repository-package` stanza, and because of `Cabal` bugs haskell/cabal#5586 and haskell/cabal#5612, running `cabal new-sdist all` causes `cabal-install` to try and package up the `source-repository-package`, leading the the "impossible happened" error. Urk.

Luckily, I think this is pretty simple to work around. Instead of doing `cabal new-sdist all`, we can simply `cd` into each project's directory and run `cabal new-sdist` individually, which is what we used to do with `cabal sdist`. I'll prepare a patch.

**Comment by phadej** [comment]
I'm confused. `haskell-ci` is "broken" because you manually add source repository packages to the script. So is `haskell-ci` broken, or your manual patch?

I'm :-1: on doing `cabal new-sdist` per package, as then there is almost no benefit over `dist`.

Right place is to fix `cabal-install` to deal with source-repositories, and for now my stance is: `haskell-ci + source-repositories = you are on your own`.

**Comment by RyanGlScott** [comment]
> So is `haskell-ci` broken, or your manual patch?

To be clear, I'm preparing a fix for #184 at the moment, but using this fix exposes the bug documented in this ticket. So `haskell-ci` is indeed broken, although it's a subtle breakage.

> I'm -1 on doing `cabal new-sdist` per package, as then there is almost no benefit over `dist`.

This isn't true. Aside from the benefits of using `all` to refer to multiple components, `new-sdist` simply works on more packages than `sdist` does. For example, I wouldn't be able to use `cabal-doctest` with GHC 7.2–based projects at all with `dist`, since their custom `Setup.hs` scripts break when combined with `dist`.

> Right place is to fix `cabal-install` to deal with source-repositories

Absolutely, I would love to see haskell/cabal#5612 fixed upstream. But that doesn't change the fact that things are broken _now_, so it would be good to have a suitable workaround until the `cabal-install` devs fix this bug.

**Comment by RyanGlScott** [comment]
Now that https://launchpad.net/~hvr/+archive/ubuntu/ghc ships `cabal-install-2.4.0.1`, which incorporates the fix for haskell/cabal#5612, this is no longer an issue. At this point, the only obstacle to using `source-repository-package`s in `haskell-ci` is #184.
